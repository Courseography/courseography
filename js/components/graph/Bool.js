import React from "react"
import { refLookUp } from "../common/utils"
import PropTypes from "prop-types"

/** Class representing a boolean node (and/or) */
export default class Bool extends React.Component {
  /**
   * Create a boolean node.
   * @param {string} state - The status of the node which can be active, inactive, or missing
   */
  constructor(props) {
    super(props)
    this.state = { status: "inactive" }
  }

  /**
   * Check whether the Bool is selected.
   * @returns {boolean} Whether status is active or not.
   */
  isSelected = () => {
    return this.state.status == "active"
  }

  /**
   * Check if the prerequisite courses have been satisfied based on bool type.
   * @returns {boolean} Whether any of the prereqs are satisfied.
   */
  arePrereqsSatisfied = () => {
    var svg = this.props.svg
    function isAllTrue(element) {
      return svg.nodes.current[element]
        ? svg.nodes.current[element].isSelected()
        : svg.bools.current[element].isSelected()
    }

    if (this.props.logicalType === "and") {
      return this.props.parents.every(isAllTrue)
    } else if (this.props.logicalType === "or") {
      return this.props.parents.some(isAllTrue)
    }
  }

  /**
   * Update the Bool's state at any moment given the prereqs and current state.
   */
  updateNode = () => {
    var svg = this.props.svg
    var newState = this.arePrereqsSatisfied() ? "active" : "inactive"

    var boolId = this.props.JSON.id_
    this.setState({ status: newState }, function () {
      localStorage.setItem(boolId, newState)
      this.props.childs.forEach(function (node) {
        var currentNode = refLookUp(node, svg)
        currentNode.updateNode(svg)
      })
      var allEdges = this.props.outEdges.concat(this.props.inEdges)
      allEdges.forEach(function (edge) {
        var currentEdge = svg.edges.current[edge]
        currentEdge.updateStatus()
      })
    })
  }

  /**
   * Cross check with the selected focus prerequisites.
   */
  focusPrereqs = () => {
    var svg = this.props.svg
    // Check if there are any missing prerequisites.
    if (this.state.status !== "active") {
      this.setState({ status: "missing" }, () => {
        this.props.inEdges.forEach(edge => {
          var currentEdge = svg.edges.current[edge]
          var sourceNode = refLookUp(currentEdge.props.source, svg)
          if (!sourceNode.isSelected()) {
            currentEdge.setState({ status: "missing" })
          }
        })
        this.props.parents.forEach(node => {
          var currentNode = refLookUp(node, svg)
          currentNode.focusPrereqs()
        })
      })
    }
  }

  /**
   * Remove the focus preqrequisites if the focus is unselected.
   */
  unfocusPrereqs = () => {
    var svg = this.props.svg
    this.updateNode(svg)
    this.props.parents.forEach(function (node) {
      var currentNode = refLookUp(node, svg)
      currentNode.unfocusPrereqs(svg)
    })
  }

  render() {
    var ellipseAttrs = {
      cx: this.props.JSON.pos[0],
      cy: this.props.JSON.pos[1],
      rx: "9.8800001",
      ry: "7.3684001",
    }
    return (
      <g
        {...this.props.JSON}
        className={this.props.className + " " + this.state.status}
        data-testid={`and(${this.props.parents.join()})`}
      >
        <ellipse {...ellipseAttrs} />
        {this.props.JSON.text.map(
          function (textTag, i) {
            var textAttrs = {
              x: ellipseAttrs.cx,
              y: textTag.pos[1],
            }
            return (
              <text {...textAttrs} key={i}>
                {this.props.logicalType}
              </text>
            )
          }.bind(this)
        )}
      </g>
    )
  }
}

Bool.propTypes = {
  childs: PropTypes.array,
  className: PropTypes.string,
  JSON: PropTypes.object,
  inEdges: PropTypes.array,
  logicalType: PropTypes.string,
  outEdges: PropTypes.array,
  parents: PropTypes.array,
  svg: PropTypes.object,
}
