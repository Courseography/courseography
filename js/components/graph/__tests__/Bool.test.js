import React from "react"
import { shallow } from "enzyme"
import Bool from "../Bool"
import { fireEvent } from "@testing-library/react"
import TestGraph from "./TestGraph"

describe("Bool", () => {
  it("should already have two classes when instantated by Graph", async () => {
    const graph = await TestGraph.build()
    const andBool = graph.getNodeByText("and")
    const orBool = graph.getNodeByText("or")

    expect(andBool.classList[0]).toBe("bool")
    expect(andBool.classList[1]).toBe("inactive")
    expect(orBool.classList[0]).toBe("bool")
    expect(orBool.classList[1]).toBe("inactive")
  })
})

describe("AND Bool", () => {
  it("should match shallow snapshot", () => {
    const boolProps = {
      JSON: {
        fill: "",
        graph: 1,
        height: 14.7368002,
        id: "bool1",
        pos: [1020.015148, 243.51310800000002],
        stroke: "",
        text: [
          {
            align: "begin",
            fill: "",
            graph: 1,
            pos: [1012.903868, 246.655709],
            rID: "text512",
            text: "and",
          },
        ],
        type_: "BoolNode",
        width: 19.7600002,
      },
      childs: ["csc369", "csc358", "csc458", "csc385", "csc367"],
      className: "bool",
      inEdges: ["p39", "p40"],
      logicalType: "and",
      outEdges: ["p41", "p42", "p43", "p59", "p82"],
      parents: ["csc209", "csc258"],
      svg: {},
    }
    const wrapper = shallow(<Bool {...boolProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it("should not do anything when you hover or click on it", async () => {
    const graph = await TestGraph.build()
    const andBool = graph.getNodeByText("and")

    expect(andBool.classList.contains("inactive")).toBe(true)

    fireEvent.mouseOver(andBool)
    expect(andBool.classList.contains("inactive")).toBe(true)

    fireEvent.click(andBool)
    expect(andBool.classList.contains("inactive")).toBe(true)
  })
  it("AND should become selected when its prereq parents are satisfied", async () => {
    const graph = await TestGraph.build()
    const andBool = graph.getNodeByText("and")
    const aaa102 = graph.getByTestId("aaa102")
    const aaa201 = graph.getByTestId("aaa201")
    const aaa303 = graph.getByTestId("aaa303")

    // AAA201 and AAA102 => AAA303
    expect(andBool.classList.contains("inactive")).toBe(true)
    fireEvent.click(aaa102)
    expect(andBool.classList.contains("inactive")).toBe(true)
    fireEvent.click(aaa201)
    expect(andBool.classList.contains("active")).toBe(true)
    expect(aaa303.classList.contains("takeable")).toBe(true)
  })
  it("AND should have the missing class when the mouse is hovering over a child class.", async () => {
    const graph = await TestGraph.build()
    const andBool = graph.getNodeByText("and")
    const aaa303 = graph.getByTestId("aaa303")

    // AAA201 and AAA102 => AAA303
    expect(andBool.classList.contains("inactive")).toBe(true)

    // mouseout or clicking triggers code to set the CSS class
    fireEvent.mouseOver(aaa303)
    fireEvent.mouseOut(aaa303)
    fireEvent.mouseOver(aaa303)
    expect(andBool.classList.contains("missing")).toBe(true)
  })
})

describe("OR Bool", () => {
  it("should match shallow snapshot", () => {
    const boolProps = {
      JSON: {
        fill: "",
        graph: 1,
        height: 14.7368002,
        id: "bool1",
        pos: [1020.015148, 243.51310800000002],
        stroke: "",
        text: [
          {
            align: "begin",
            fill: "",
            graph: 1,
            pos: [1012.903868, 246.655709],
            rID: "text512",
            text: "or",
          },
        ],
        type_: "BoolNode",
        width: 19.7600002,
      },
      childs: ["csc369", "csc358", "csc458", "csc385", "csc367"],
      className: "bool",
      inEdges: ["p39", "p40"],
      logicalType: "or",
      outEdges: ["p41", "p42", "p43", "p59", "p82"],
      parents: ["csc209", "csc258"],
      svg: {},
    }
    const wrapper = shallow(<Bool {...boolProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it("should not do anything when you hover or click on it", async () => {
    const graph = await TestGraph.build()
    const orBool = graph.getNodeByText("or")

    expect(orBool.classList.contains("inactive")).toBe(true)

    fireEvent.mouseOver(orBool)
    expect(orBool.classList.contains("inactive")).toBe(true)

    fireEvent.click(orBool)
    expect(orBool.classList.contains("inactive")).toBe(true)
  })
  it("or should become selected when its prereq parents are satisfied", async () => {
    const graph = await TestGraph.build()
    const orBool = graph.getNodeByText("or")
    const aaa102 = graph.getByTestId("aaa102")
    const aaa201 = graph.getByTestId("aaa201")
    const aaa202 = graph.getByTestId("aaa202")

    // AAA201 or AAA102 => AAA202
    expect(orBool.classList.contains("inactive")).toBe(true)

    fireEvent.click(aaa102)
    expect(orBool.classList.contains("active")).toBe(true)
    expect(aaa202.classList.contains("takeable")).toBe(true)

    fireEvent.click(aaa201)
    expect(orBool.classList.contains("active")).toBe(true)
    expect(aaa202.classList.contains("takeable")).toBe(true)
  })
  it("or should have the missing class when the mouse is hovering over a child class.", async () => {
    const graph = await TestGraph.build()
    const orBool = graph.getNodeByText("or")
    const aaa202 = graph.getByTestId("aaa202")

    // AAA201 or AAA102 => AAA202
    expect(orBool.classList.contains("inactive")).toBe(true)
    fireEvent.mouseOver(aaa202)
    fireEvent.mouseOut(aaa202)
    fireEvent.mouseOver(aaa202)
    expect(orBool.classList.contains("missing")).toBe(true)
  })
})
