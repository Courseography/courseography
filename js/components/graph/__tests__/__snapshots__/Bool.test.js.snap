// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AND Bool should match shallow snapshot 1`] = `
<g
  className="bool inactive"
  data-testid="and(csc209,csc258)"
  fill=""
  graph={1}
  height={14.7368002}
  id="bool1"
  pos={
    Array [
      1020.015148,
      243.51310800000002,
    ]
  }
  stroke=""
  text={
    Array [
      Object {
        "align": "begin",
        "fill": "",
        "graph": 1,
        "pos": Array [
          1012.903868,
          246.655709,
        ],
        "rID": "text512",
        "text": "and",
      },
    ]
  }
  type_="BoolNode"
  width={19.7600002}
>
  <ellipse
    cx={1020.015148}
    cy={243.51310800000002}
    rx="9.8800001"
    ry="7.3684001"
  />
  <text
    key="0"
    x={1020.015148}
    y={246.655709}
  >
    and
  </text>
</g>
`;

exports[`OR Bool should match shallow snapshot 1`] = `
<g
  className="bool inactive"
  data-testid="and(csc209,csc258)"
  fill=""
  graph={1}
  height={14.7368002}
  id="bool1"
  pos={
    Array [
      1020.015148,
      243.51310800000002,
    ]
  }
  stroke=""
  text={
    Array [
      Object {
        "align": "begin",
        "fill": "",
        "graph": 1,
        "pos": Array [
          1012.903868,
          246.655709,
        ],
        "rID": "text512",
        "text": "or",
      },
    ]
  }
  type_="BoolNode"
  width={19.7600002}
>
  <ellipse
    cx={1020.015148}
    cy={243.51310800000002}
    rx="9.8800001"
    ry="7.3684001"
  />
  <text
    key="0"
    x={1020.015148}
    y={246.655709}
  >
    or
  </text>
</g>
`;
